core/trino-main/src/main/java/io/trino/cost/TopNRankingStatsRule.java

```diff
@@ -0,0 +1,218 @@
+            // at most (maxRank * rowsPerDistinctValue). In the above case rowsPerDistinctValue is 2 based on distinct
+            // count. Hence, the number of rows per partition for max rank 2 should be 4. Whereas, the distinct count
+            // for rank symbol should be 2 i.e. maxRank.
+            double distinctCount = estimateCorrelatedDistinctCount(node.getOrderingScheme().orderBy(), sourceStats);
+            // This should be >= 1.0, otherwise we will underestimate the rows per partition.
+            double rowsPerDistinctValue = ceil(maxRowsPerPartition / distinctCount);
+            rowsPerPartition = min(maxRowsPerPartition, node.getMaxRankingPerPartition() * rowsPerDistinctValue);
+            rankDistinctCount = min(distinctCount, node.getMaxRankingPerPartition());
+        }
+
+        if (isNaN(rowsPerPartition)) {
+            rowsPerPartition = maxRowsPerPartition;
+        }
+
+        if (isNaN(rankDistinctCount)) {
+            rankDistinctCount = rowsPerPartition;
+        }
+
+        PlanNodeStatsEstimate.Builder estimateBuilder = PlanNodeStatsEstimate.buildFrom(sourceStats);
+        fixOrderBySymbolDistinctCount(
+                node.getOrderingScheme().orderBy(),
+                partitionCount,
+                rowsPerPartition,
+                maxRowsPerPartition,
+                sourceStats,
+                estimateBuilder);
+
+        double outputRowsCount = partitionCount * rowsPerPartition;
+        return Optional.of(estimateBuilder
+                .setOutputRowCount(outputRowsCount)
+                .addSymbolStatistics(node.getRankingSymbol(), SymbolStatsEstimate.builder()
+                        .setLowValue(1)
+                        // maxRankingPerPartition is the maximum rank per partition. Therefore, in the worse case
+                        // scenario the rank symbol can have maxRankingPerPartition as the high value.
+                        .setHighValue(node.getMaxRankingPerPartition())
+                        .setDistinctValuesCount(rankDistinctCount)
+                        .setNullsFraction(0.0)
+                        .setAverageRowSize(BIGINT.getFixedSize())
+                        .build())
+                .build());
+    }
+
+    private static void fixOrderBySymbolDistinctCount(
+            List<Symbol> orderBy,
+            double partitionCount,
+            double rowsPerPartition,
+            double maxRowsPerPartition,
+            PlanNodeStatsEstimate sourceStats,
+            PlanNodeStatsEstimate.Builder estimateBuilder)
+    {
+        if (orderBy.isEmpty()) {
+            return;
+        }
+        Symbol firstSortSymbol = orderBy.getFirst();
+        SymbolStatsEstimate symbolStats = sourceStats.getSymbolStatistics(firstSortSymbol);
+        double newDistinctCountPerPartition = ceil((symbolStats.getDistinctValuesCount() / maxRowsPerPartition) * rowsPerPartition);
```

Comment:
it's not `newDistinctCountPerPartition`, but rather `newDistinctCount`. It's across all partitions (global)